std/socket.zig

```diff
@@ -0,0 +1,678 @@
+                system.ERROR.WSAEAFNOSUPPORT => return SocketError.AddressFamilyNotSupported,
+                system.ERROR.WSAEINVAL => return SocketError.ProtocolFamilyNotAvailable,
+                system.ERROR.WSAEMFILE => return SocketError.ProcessFdQuotaExceeded,
+                system.ERROR.WSAENOBUFS => return SocketError.system.emResources,
+                system.ERROR.WSAEPROTONOSUPPORT, system.ERROR.WSAEPROTOTYPE, system.ERROR.WSAESOCKTNOSUPPORT => return SocketError.ProtocolNotSupported,
+                else => return unexpectedError(@intCast(u32, err)),
+            }
+        }
+    }
+
+    pub fn tcp(domain: Domain) SocketError!Socket {
+        return Socket.new(domain, SocketType.Stream, Protocol.TCP);
+    }
+
+    pub fn udp(domain: Domain) SocketError!Socket {
+        return Socket.new(domain, SocketType.DataGram, Protocol.UDP);
+    }
+
+    pub fn unix(socket_type: SocketType) SocketError!Socket {
+        return Socket.new(Domain.Unix, socket_type, Protocol.Unix);
+    }
+
+    pub fn bind(self: Socket, addr: *const net.Address) BindError!void {
+        const rc = system.bind(self.fd, &addr.os_addr, @sizeOf(net.OsAddress));
+
+        if (is_posix) {
+            const err = system.getErrno(@intCast(usize, rc));
+            switch (err) {
+                0 => return,
+                system.EACCES => return BindError.AccessDenied,
+                system.EADDRINUSE => return BindError.AddressInUse,
+                system.EBADF => unreachable, // always a race condition if this error is returned
+                system.EINVAL => unreachable,
+                system.ENOTSOCK => unreachable,
+                system.EADDRNOTAVAIL => return BindError.AddressNotAvailable,
+                system.EFAULT => unreachable,
+                system.ELOOP => return BindError.SymLinkLoop,
+                system.ENAMETOOLONG => return BindError.NameTooLong,
+                system.ENOENT => return BindError.FileNotFound,
+                system.ENOMEM => return BindError.system.emResources,
+                system.ENOTDIR => return BindError.NotDir,
+                system.EROFS => return BindError.ReadOnlyFilesystem.em,
+                else => return unexpectedError(err),
+            }
+        } else {
+            const err = system.WSAGetLastError();
+            switch (err) {
+                0 => return,
+                system.ERROR.WSAEACCES => return BindError.AccessDenied,
+                system.ERROR.WSAEADDRINUSE => return BindError.AddressInUse,
+                system.ERROR.WSAEINVAL => unreachable,
+                system.ERROR.WSAENOTSOCK => unreachable,
+                system.ERROR.WSAEADDRNOTAVAIL => return BindError.AddressNotAvailable,
+                system.ERROR.WSAEFAULT => unreachable,
+                else => return unexpectedError(@intCast(u32, err)),
+            }
+        }
+    }
+
+    pub fn listen(self: Socket, backlog: u32) ListenError!void {
```

Comment:
Most systems these days ignore the `backlog` argument. Maybe we shouldn't include it?

Reply:
We could have `listen` and forward it to `listenBacklog`?